tfx/components/example_gen/utils.py

```diff
@@ -209,84 +213,183 @@ def _glob_to_regex(glob_pattern: Text) -> Text:
+      latest_version = select_version
+    if latest_version is None or version >= int(latest_version):
+      latest_version = select_version
 
-  if latest_span is None:
-    raise ValueError('Cannot not find matching for split %s based on %s' %
+  if latest_span is None or (is_match_version and latest_version is None):
+    raise ValueError('Cannot find matching for split %s based on %s' %
                      (split.name, split.pattern))
-  return latest_span
+
+  return latest_span, latest_version
+
+
+def _retrieve_latest_span_version(uri: Text,
+                                  split: example_gen_pb2.Input.Split
+                                  ) -> Tuple[Optional[Text], Optional[Text]]:
+  """Retrieves the most recent span and version for a given split pattern.
+
+  If both Span and Version spec occur in the split pattern, searches for and
+  returns both the latest Span and Version. If only Span exists in the split
+  pattern, searches for the latest Span, and Version is returned as None.
+  If Version is present, but not Span, an error is raised. If neither Span
+  nor Version is present, returns both as None.
+
+  Args:
+    uri: The base path from which files will be searched.
+    split: An example_gen_pb2.Input.Split object which contains a split pattern,
+      to be searched on. Note that this function will update the {SPAN} in this
+      and {VERSION} tags in the split config to actual Span and Version numbers.
+  Returns:
+    Tuple of two strings, Span (optional) and Version (optional).
+
+  Raises:
+    ValueError: if any of the following occurs:
+      - If either Span or Version spec is occurs in the split pattern
+        more than once.
+      - If Version spec is provided, but Span spec is not present.
+      - If Span or Version found is not an integer.
+      - If a matching cannot be found for split pattern provided.
+  """
+
+  latest_span = None
+  latest_version = None
+
+  if SPAN_SPEC in split.pattern:
+    if split.pattern.count(SPAN_SPEC) != 1:
+      raise ValueError('Only one %s is allowed in %s' % (SPAN_SPEC,
+                                                         split.pattern))
+
+    is_match_version = VERSION_SPEC in split.pattern
+    if is_match_version and split.pattern.count(VERSION_SPEC) != 1:
+      raise ValueError('Only one %s is allowed in %s' % (VERSION_SPEC,
+                                                           split.pattern))
+
+    latest_span, latest_version = _glob_and_match_span_version(uri, split,
+                                                               is_match_version)
+
+  elif VERSION_SPEC in split.pattern:
+    raise ValueError('Version spec provided, but Span spec is not present')
+
+  # Replace split.pattern so executor can find files after driver runs.
+  if latest_span:
+    split.pattern = split.pattern.replace(SPAN_SPEC, latest_span)
+  if latest_version:
+    split.pattern = split.pattern.replace(VERSION_SPEC, latest_version)
```

Comment:
I feel we can move this to calculate_splits_fingerprint_span_and_version, and keep this function as a read only function?

Reply:
moved this to `calculate_splits_fp_span_and_ver4sion()`